README
README - info file
Copyright (C) 2005 Free Software Foundation, Inc.
Written by Maurizio Boriani <baux@member.fsf.org>

This file is part of the greg.
 
The greg is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
 
The greg is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
 
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA


* About

 Greg is a framework for testing other programs and libraries.
 Its purpose is to provide a single front end for all tests and to
 be a small, simple framework for writing tests.
 Greg leverages off the Guile language to provide all the power (and more)
 of other test frameworks with greater simplicity and ease of use.

 Please send bug reports to https://gna.org/bugs/?group=greg or
 greg-devel@gna.org

* greg dependencies

 greg depends from this packages, need them to build and run:
	
	Guile: http://www.gnu.org/software/guile/  
		The guile scheme implementation
	libc, gcc and so on...
 
* greg Documentation

 The doc directory contains the reference manual  and some examples.

 The www pages are at http://www.gnu.org/software/greg or
 http://home.gna.org/greg/

 The project page is at: http://www.gna.org/projects/greg.

* Development versions

 greg source code is mantained using gnu arch. Main archives mirror
 can be found at:

 archive: baux@member.fsf.org--2005-greg
 location: http://arch.gna.org/upload/greg/2005-greg

 branches: 
    greg--baux   (my personal brench)
    greg--devel  (development, tag to this if you'd like to 
      contribute)
    greg--stable (released versions).
    greg-wpages  (web pages).

 Have a look at gnu arch manuals.    

* greg License

 The greg main script is under GPL License while lib is under LGPL.

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2, or (at your option)
 any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this software; see the file COPYING.  If not, write to
 the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 Boston, MA 02111-1307 USA


NEWS
greg NEWS --- history of user-visible changes.
Copyright (C) 2005 Free Software Foundation, Inc.
See the end for copying conditions.

Please send greg bug reports to greg-devel@gna.org or 

https://gna.org/bugs/?group=greg-dbi



Changes in greg 1.5.5:

* Ugraded gnu toolchain
* Fixed guile api use to more recent releases (and 
  back compatible with older).


test/test.sh
#!/bin/sh

# test.sh -- prepare the env and execute a self test
# Copyright (C) 1998, 2000, 2001, 2003, 2005 Free Software Foundation, Inc.
# Written by Maurizio Boriani <baux@member.fsf.org>
#
# This file is part of the greg.
# 
# The greg is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# The greg is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
#
#                                               -*- Makefile -*-
# Process this file with autoconf to produce a configure script.

export LTDL_LIBRARY_PATH=$LTDL_LIBRARY_PATH:../src/.libs
export GUILE_LOAD_PATH=$GUILE_LOAD_PATH:../src/
guile -s ../src/gregtool


test/basic/greg.scm

;   Copyright (C) 1998 Free Software Foundation, Inc.
;
;   Written by:  Richard Frith-Macdonald <rfm@gnu.org>
;   Date: April 1998
;
; This file is part of the Greg package - part of the GNUstep project.
;
; This library is free software; you can redistribute it and/or
; modify it under the terms of the GNU Lesser General Public
; License as published by the Free Software Foundation; either
; version 2.1 of the License, or (at your option) any later version.
; 
; This library is distributed in the hope that it will be useful,
; but WITHOUT ANY WARRANTY; without even the implied warranty of
; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
; Lesser General Public License for more details.
; 
; You should have received a copy of the GNU Lesser General Public
; License along with this library; if not, write to the Free Software
; Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
;

;
;	Test that all possible exits from greg-testcase work as expected.
;

(use-modules (greg))

(greg-testcase "We can return an expected pass" #t
(lambda ()
  (greg-recv ("^PASS: We can return an expected pass" #t))
))

(greg-testcase "We can return an expected fail" #t
(lambda ()
  (greg-recv ("^XFAIL: We can return an expected fail" #t))
))

(greg-testcase "We can return an unexpected pass" #t
(lambda ()
  (greg-recv ("^UPASS: We can return an unexpected pass" #t))
))

(greg-testcase "We can return an unexpected fail" #t
(lambda ()
  (greg-recv ("^FAIL: We can return an unexpected fail" #t))
))

(greg-testcase "We can return an unresolved" #t
(lambda ()
  (greg-recv ("^UNRESOLVED: We can return an unresolved" #t))
))

(greg-testcase "We can throw an expected pass" #t
(lambda ()
  (greg-recv ("^PASS: We can throw an expected pass" #t))
))

(greg-testcase "We can throw an expected fail" #t
(lambda ()
  (greg-recv ("^XFAIL: We can throw an expected fail" #t))
))

(greg-testcase "We can throw an unexpected pass" #t
(lambda ()
  (greg-recv ("^UPASS: We can throw an unexpected pass" #t))
))

(greg-testcase "We can throw an unexpected fail" #t
(lambda ()
  (greg-recv ("^FAIL: We can throw an unexpected fail" #t))
))

(greg-testcase "We can throw an untested" #t
(lambda ()
  (greg-recv ("^UNTESTED: We can throw an untested" #t))
))

(greg-testcase "We can throw an unsupported" #t
(lambda ()
  (greg-recv ("^UNSUPPORTED: We can throw an unsupported" #t))
))

(greg-testcase "We can throw an unresolved" #t
(lambda ()
  (greg-recv ("^UNRESOLVED: We can throw an unresolved" #t))
))

;
;	Now do it all again with strict posix
;
(set! greg-posix #t)

(greg-testcase "We can return an expected pass" #t
(lambda ()
  (greg-recv ("^PASS: We can return an expected pass" #t))
))

(greg-testcase "We can't return an expected fail" #t
(lambda ()
  (greg-recv ("^FAIL: We can't return an expected fail" #t))
))

(greg-testcase "We can't return an unexpected pass" #t
(lambda ()
  (greg-recv ("^PASS: We can't return an unexpected pass" #t))
))

(greg-testcase "We can return an unexpected fail" #t
(lambda ()
  (greg-recv ("^FAIL: We can return an unexpected fail" #t))
))

(greg-testcase "We can return an unresolved" #t
(lambda ()
  (greg-recv ("^UNRESOLVED: We can return an unresolved" #t))
))

(greg-testcase "We can throw an expected pass" #t
(lambda ()
  (greg-recv ("^PASS: We can throw an expected pass" #t))
))

(greg-testcase "We can't throw an expected fail" #t
(lambda ()
  (greg-recv ("^FAIL: We can't throw an expected fail" #t))
))

(greg-testcase "We can't throw an unexpected pass" #t
(lambda ()
  (greg-recv ("^PASS: We can't throw an unexpected pass" #t))
))

(greg-testcase "We can throw an unexpected fail" #t
(lambda ()
  (greg-recv ("^FAIL: We can throw an unexpected fail" #t))
))

(greg-testcase "We can throw an untested" #t
(lambda ()
  (greg-recv ("^UNTESTED: We can throw an untested" #t))
))

(greg-testcase "We can throw an unsupported" #t
(lambda ()
  (greg-recv ("^UNSUPPORTED: We can throw an unsupported" #t))
))

(greg-testcase "We can throw an unresolved" #t
(lambda ()
  (greg-recv ("^UNRESOLVED: We can throw an unresolved" #t))
))



test/tests/greg.scm

;   Copyright (C) 1998 Free Software Foundation, Inc.
;
;   Written by:  Richard Frith-Macdonald <rfm@gnu.org>
;   Date: April 1998
;
; This file is part of the Greg package - part of the GNUstep project.
;
; This library is free software; you can redistribute it and/or
; modify it under the terms of the GNU Lesser General Public
; License as published by the Free Software Foundation; either
; version 2.1 of the License, or (at your option) any later version.
; 
; This library is distributed in the hope that it will be useful,
; but WITHOUT ANY WARRANTY; without even the implied warranty of
; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
; Lesser General Public License for more details.
; 
; You should have received a copy of the GNU Lesser General Public
; License along with this library; if not, write to the Free Software
; Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
;

(use-modules (greg))

(greg-testcase "We can return an expected pass" #t
(lambda ()
  #t
))


(greg-testcase "We can return an expected fail" #f
(lambda ()
  #f
))


(greg-testcase "We can return an unexpected pass" #f
(lambda ()
  #t
))


(greg-testcase "We can return an unexpected fail" #t
(lambda ()
  #f
))


(greg-testcase "We can return an unresolved" #t
(lambda ()
  "xxx"
))


(greg-testcase "We can throw an expected pass" #t
(lambda ()
  (throw 'pass)
))


(greg-testcase "We can throw an expected fail" #f
(lambda ()
  (throw 'fail)
))


(greg-testcase "We can throw an unexpected pass" #f
(lambda ()
  (throw 'pass)
))


(greg-testcase "We can throw an unexpected fail" #t
(lambda ()
  (throw 'fail)
))


(greg-testcase "We can throw an untested" #t
(lambda ()
  (throw 'untested)
))


(greg-testcase "We can throw an unsupported" #t
(lambda ()
  (throw 'unsupported)
))


(greg-testcase "We can throw an unresolved" #t
(lambda ()
  (throw 'unresolved)
))


(set! greg-posix #t)

(greg-testcase "We can return an expected pass" #t
(lambda ()
  #t
))


(greg-testcase "We can't return an expected fail" #f
(lambda ()
  #f
))


(greg-testcase "We can't return an unexpected pass" #f
(lambda ()
  #t
))


(greg-testcase "We can return an unexpected fail" #t
(lambda ()
  #f
))


(greg-testcase "We can return an unresolved" #t
(lambda ()
  "xxx"
))


(greg-testcase "We can throw an expected pass" #t
(lambda ()
  (throw 'pass)
))


(greg-testcase "We can't throw an expected fail" #f
(lambda ()
  (throw 'fail)
))


(greg-testcase "We can't throw an unexpected pass" #f
(lambda ()
  (throw 'pass)
))


(greg-testcase "We can throw an unexpected fail" #t
(lambda ()
  (throw 'fail)
))


(greg-testcase "We can throw an untested" #t
(lambda ()
  (throw 'untested)
))


(greg-testcase "We can throw an unsupported" #t
(lambda ()
  (throw 'unsupported)
))


(greg-testcase "We can throw an unresolved" #t
(lambda ()
  (throw 'unresolved)
))



src/greg.c
/* greg.c
   Copyright (C) 1998 Free Software Foundation, Inc.
   
   Written by:  Richard frith-Macdonald <rfm@gnu.org>
   Date: 1998
   
   This file is part of the Greg package - part of the GNUstep project.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.
   
   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.
   
   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
   */ 

#include "config.h"

#if	HAVE_SIGNAL_H
#include <signal.h>
#endif
#if	HAVE_SYS_FILE_H
#include <sys/file.h>
#endif
#if	HAVE_SYS_FCNTL_H
#include <sys/fcntl.h>
#endif
#if	HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif
#if	HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif
#if HAVE_ERRNO_H
#include <errno.h>
#else
extern int	errno;
#endif

#ifdef __MINGW32__
#define __GUILE_IMPORT__ 1
#else
#ifdef __WIN32__
#define __GUILE_IMPORT__ 1
#endif
#endif

#include <libguile.h>

/*
 *	If we are on a streams based system, we need to include stropts.h
 *	for definitions needed to set up slave pseudo-terminal stream.
 */
#if	HAVE_SYS_STROPTS_H
#include <sys/stropts.h>
#endif

#ifndef	MAX_OPEN
#define	MAX_OPEN	64
#endif

static int
pty_master(char* name, int len)
{
  int	master;

  /*
   *	If we have grantpt(), assume we are using sysv-style pseudo-terminals,
   *	otherwise assume bsd style.
   */
#if	HAVE_GRANTPT
  master = open("/dev/ptmx", O_RDWR);
  if (master >= 0)
    {
      const char	*slave;

      grantpt(master);                   /* Change permission of slave.  */
      unlockpt(master);                  /* Unlock slave.        */
      slave = (const char*)ptsname(master);
      if (slave == 0 || strlen(slave) >= len)
	{
	  close(master);
	  master = -1;
	}
      else
	{
	  strcpy(name, (char*)slave);
	}
    }
  else
#endif
    {
      const char	*groups = "pqrstuvwxyzPQRSTUVWXYZ";

      master = -1;
      if (len > 10)
        {
	  strcpy(name, "/dev/ptyXX");
	  while (master < 0 && *groups != '\0')
	    {
	      int	i;

	      name[8] = *groups++;
	      for (i = 0; i < 16; i++)
	        {
		  name[9] = "0123456789abcdef"[i];
		  master = open(name, O_RDWR);
		  if (master >= 0)
		    {
		      name[5] = 't';
		      break;
		    }
		}
	    }
	}
    }
  return master;
}

static int
pty_slave(const char* name)
{
  int	slave;

  slave = open(name, O_RDWR);
#if	HAVE_SYS_STROPTS_H
#if	HAVE_PTS_STREAM_MODULES
  if (slave >= 0 && isastream(slave))
    {
      if (ioctl(slave, I_PUSH, "ptem") < 0)
	{
	  perror("unable to push 'ptem' streams module");
	}
      else if (ioctl(slave, I_PUSH, "ldterm") < 0)
	{
	  perror("unable to push 'ldterm' streams module");
	}
    }
#endif
#endif
  return slave;
}

static char	s_pty_child[] = "pty-child";

SCM_DEFINE (scm_pty_child, "pty-child", 0, 0, 1,
	    (SCM args),
	    "Open a pty child.")
#define FUNC_NAME s_pty_child
{
  SCM	ans = SCM_EOL;
  SCM	prg;
  char	slave_name[32];
  int	master;

  prg = SCM_CAR(args);
  /*
   *	We permit this to be called either with multiple string arguments
   *	or with a list of arguments.
   */
  while (scm_list_p(prg) == SCM_BOOL_T && SCM_CDR(args) == SCM_EOL)
    {
      args = prg;
      prg = SCM_CAR(args);
    }
  SCM_ASSERT (SCM_NIMP(prg) && scm_is_string(prg), prg, SCM_ARG1, s_pty_child);

  master = pty_master(slave_name, sizeof(slave_name));
  if (master >= 0)
    {
      int	p[2];
      int	pid;
      if (pipe(p) < 0)
	{
	  (void)close(master);
	  scm_misc_error("pty-child", "failed to open pipe", SCM_EOL);
	}
      pid = fork();
      if (pid < 0)
	{
	  (void)close(master);
	  (void)close(p[0]);
	  (void)close(p[1]);
	  scm_misc_error("pty-child", "failed to fork child pipe", SCM_EOL);
	}
      if (pid == 0)
	{
	  int	i;
	  int	s;

	  for (i = 1; i < 32; i++)
	    {
	      signal(i, SIG_DFL);
	    }

	  s = pty_slave(slave_name);
	  if (s < 0)
	    {
	      char		buf[256];
	      int		len;
#if	HAVE_STRERROR
	      const char	*msg = strerror(errno);
#else
	      char		msg[32];

	      sprintf(msg, "system error %d", errno);
#endif

	      len = strlen(msg);
	      if (len > 255)
		len = 255;
	      buf[0] = len;
	      memcpy(&buf[1], msg, len);
	      write(p[1], buf, len+1);	/* Tell parent we failed.	*/
	      exit(1);			/* Failed to open slave!	*/
	    }

	  for (i = 0; i < MAX_OPEN; i++)
	    {
	      if (i != s && i != 2 && i != p[1])
		{
		  (void)close(i);
		}
	    }

	  i = -1;
#if	HAVE_SETSID
	  i = setsid();
#endif
#if	HAVE_SETPGID
	  if (i < 0)
	    {
	      i = getpid();
	      i = setpgid(i, i);
	    }
#endif
#ifdef	TIOCNOTTY
	  i = open("/dev/tty", O_RDWR);
	  if (i >= 0)
	    {
	      (void)ioctl(i, TIOCNOTTY, 0);
	      (void)close(i);
	    }
#endif

	  /*
	   * Make sure pty operates on stdin and stdout.
	   * If the pipe to our parent is on stdin or stdout, duplicate
	   * it so that it doesn't get closed prematurely.
           */
	  if (s != 0)
	    {
	      if (p[1] == 0)
		p[1] = dup(p[1]);
	      (void)dup2(s, 0);
	    }
	  if (s != 1)
	    {
	      if (p[1] == 1)
		p[1] = dup(p[1]);
	      (void)dup2(s, 1);
	    }

	  /*
	   * Discard original descriptor - we no longer need it.
	   */
	  if (s > 1)
	    {
	      (void)close(s);
	    }

	  (void)write(p[1], "", 1);	/* Tell parent we are ready.	*/
	  (void)close(p[1]);
	  (void)dup2(1, 2);		/* Send stderr to pty.		*/

	  if (scm_string_equal_p(prg, gh_str02scm("")) != SCM_BOOL_T)
	    {
	      scm_execl(prg, args);
	      exit(1);
	    }
	  else
	    {
	      /*
	       *	Program name is an empty string - don't exec a
	       *	child, just return a list containing 0 to mark
	       *	this as a success (but in the child process).
	       */
	      ans = scm_cons(SCM_MAKINUM(0), SCM_EOL);
	    }
	}
      else
	{
	  unsigned char	info;
	  int	len;
	  SCM	cpid;
	  SCM	rport;
	  SCM	wport;

	  (void)close(p[1]);
	  /*
	   *	Synchronize with child process - it should send us a byte
	   *	when everything is set up - immediately before the exec.
	   */
	  len = read(p[0], &info, 1);
	  if (len != 1)
	    {
	      (void)close(p[0]);
	      (void)close(master);
#if	HAVE_WAITPID
	      {
		int	status;
		int	opts = 0;
		(void)waitpid(pid, &status, opts);
	      }
#endif
	      scm_misc_error("pty-child", "failed to sync with child",
			SCM_EOL);
	    }
	  if (info == '\0')
	    {
	      (void)close(p[0]);
	    }
	  else
	    {
	      char	buf[1024];
	      int	base;

	      sprintf(buf, "child failed to open %s: ", slave_name);
	      base = strlen(buf);
	
	      len = read(p[0], &buf[base], (int)info);
	      if (len > 0)
		{
		  buf[base+len] = '\0';
		}
	      else
		{
		  strcat(buf, "unknown error");
		}
	      (void)close(p[0]);
	      scm_misc_error("pty-child", buf, SCM_EOL);
	    }
	  cpid = SCM_MAKINUM(pid);
	  rport = scm_fdopen(SCM_MAKINUM(master), scm_makfrom0str("r"));
	  wport = scm_fdopen(SCM_MAKINUM(master), scm_makfrom0str("w"));
	  ans = scm_cons(rport, scm_cons(wport, scm_cons(cpid, SCM_EOL)));
	}
    }
  else
    {
      scm_misc_error("pty-child", "failed to get master pty", SCM_EOL);
    }
  return ans;
}
#undef FUNC_NAME

void
scm_init_greg()
{
#ifndef SCM_MAGIC_SNARFER
#include "greg.x"
#endif
}



src/libgreg.c
/* greg.c
   Copyright (C) 1998 Free Software Foundation, Inc.
   
   Written by:  Richard frith-Macdonald <rfm@gnu.org>
   Date: 1998
   
   This file is part of the Greg package - part of the GNUstep project.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.
   
   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.
   
   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
   */ 

#include "config.h"

#if        HAVE_SIGNAL_H
#include <signal.h>
#endif
#if        HAVE_SYS_FILE_H
#include <sys/file.h>
#endif
#if        HAVE_SYS_FCNTL_H
#include <sys/fcntl.h>
#endif
#if        HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif
#if        HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif
#if HAVE_ERRNO_H
#include <errno.h>
#else
extern int        errno;
#endif

#ifdef __MINGW32__
#define __GUILE_IMPORT__ 1
#else
#ifdef __WIN32__
#define __GUILE_IMPORT__ 1
#endif
#endif

#include <libguile.h>

/*
 *        If we are on a streams based system, we need to include stropts.h
 *        for definitions needed to set up slave pseudo-terminal stream.
 */
#if        HAVE_SYS_STROPTS_H
#include <sys/stropts.h>
#endif

#ifndef        MAX_OPEN
#define        MAX_OPEN        64
#endif

static int
pty_master(char* name, int len)
{
  int        master;

  /*
   *        If we have grantpt(), assume we are using sysv-style pseudo-terminals,
   *        otherwise assume bsd style.
   */
#if        HAVE_GRANTPT
  master = open("/dev/ptmx", O_RDWR);
  if (master >= 0)
    {
      const char        *slave;

      grantpt(master);                   /* Change permission of slave.  */
      unlockpt(master);                  /* Unlock slave.        */
      slave = (const char*)ptsname(master);
      if (slave == 0 || strlen(slave) >= len)
        {
          close(master);
          master = -1;
        }
      else
        {
          strcpy(name, (char*)slave);
        }
    }
  else
#endif
    {
      const char        *groups = "pqrstuvwxyzPQRSTUVWXYZ";

      master = -1;
      if (len > 10)
        {
          strcpy(name, "/dev/ptyXX");
          while (master < 0 && *groups != '\0')
            {
              int        i;

              name[8] = *groups++;
              for (i = 0; i < 16; i++)
                {
                  name[9] = "0123456789abcdef"[i];
                  master = open(name, O_RDWR);
                  if (master >= 0)
                    {
                      name[5] = 't';
                      break;
                    }
                }
            }
        }
    }
  return master;
}

static int
pty_slave(const char* name)
{
  int        slave;

  slave = open(name, O_RDWR);
#if        HAVE_SYS_STROPTS_H
#if        HAVE_PTS_STREAM_MODULES
  if (slave >= 0 && isastream(slave))
    {
      if (ioctl(slave, I_PUSH, "ptem") < 0)
        {
          perror("unable to push 'ptem' streams module");
        }
      else if (ioctl(slave, I_PUSH, "ldterm") < 0)
        {
          perror("unable to push 'ldterm' streams module");
        }
    }
#endif
#endif
  return slave;
}

static char        s_pty_child[] = "pty-child";

SCM_DEFINE (scm_pty_child, "pty-child", 0, 0, 1,
            (SCM args),
            "Open a pty child.")
#define FUNC_NAME s_pty_child
{
  SCM        ans = SCM_EOL;
  SCM        prg;
  char        slave_name[32];
  int        master;

  prg = SCM_CAR(args);
  /*
   *        We permit this to be called either with multiple string arguments
   *        or with a list of arguments.
   */
  while (scm_list_p(prg) == SCM_BOOL_T && SCM_CDR(args) == SCM_EOL)
    {
      args = prg;
      prg = SCM_CAR(args);
    }
  SCM_ASSERT (SCM_NIMP(prg) && scm_is_string(prg), prg, SCM_ARG1, s_pty_child);

  master = pty_master(slave_name, sizeof(slave_name));
  if (master >= 0)
    {
      int        p[2];
      int        pid;
      if (pipe(p) < 0)
        {
          (void)close(master);
          scm_misc_error("pty-child", "failed to open pipe", SCM_EOL);
        }
      pid = fork();
      if (pid < 0)
        {
          (void)close(master);
          (void)close(p[0]);
          (void)close(p[1]);
          scm_misc_error("pty-child", "failed to fork child pipe", SCM_EOL);
        }
      if (pid == 0)
        {
          int        i;
          int        s;

          for (i = 1; i < 32; i++)
            {
              signal(i, SIG_DFL);
            }

          s = pty_slave(slave_name);
          if (s < 0)
            {
              char                buf[256];
              int                len;
#if        HAVE_STRERROR
              const char        *msg = strerror(errno);
#else
              char                msg[32];

              sprintf(msg, "system error %d", errno);
#endif

              len = strlen(msg);
              if (len > 255)
                len = 255;
              buf[0] = len;
              memcpy(&buf[1], msg, len);
              write(p[1], buf, len+1);        /* Tell parent we failed.        */
              exit(1);                        /* Failed to open slave!        */
            }

          for (i = 0; i < MAX_OPEN; i++)
            {
              if (i != s && i != 2 && i != p[1])
                {
                  (void)close(i);
                }
            }

          i = -1;
#if        HAVE_SETSID
          i = setsid();
#endif
#if        HAVE_SETPGID
          if (i < 0)
            {
              i = getpid();
              i = setpgid(i, i);
            }
#endif
#ifdef        TIOCNOTTY
          i = open("/dev/tty", O_RDWR);
          if (i >= 0)
            {
              (void)ioctl(i, TIOCNOTTY, 0);
              (void)close(i);
            }
#endif

          /*
           * Make sure pty operates on stdin and stdout.
           * If the pipe to our parent is on stdin or stdout, duplicate
           * it so that it doesn't get closed prematurely.
           */
          if (s != 0)
            {
              if (p[1] == 0)
                p[1] = dup(p[1]);
              (void)dup2(s, 0);
            }
          if (s != 1)
            {
              if (p[1] == 1)
                p[1] = dup(p[1]);
              (void)dup2(s, 1);
            }

          /*
           * Discard original descriptor - we no longer need it.
           */
          if (s > 1)
            {
              (void)close(s);
            }

          (void)write(p[1], "", 1);        /* Tell parent we are ready.        */
          (void)close(p[1]);
          (void)dup2(1, 2);                /* Send stderr to pty.                */

          if (scm_string_null_p(prg) != SCM_BOOL_T)
            {
              scm_execl(prg, args);
              exit(1);
            }
          else
            {
              /*
               *        Program name is an empty string - don't exec a
               *        child, just return a list containing 0 to mark
               *        this as a success (but in the child process).
               */
              ans = scm_cons(scm_from_int(0), SCM_EOL);
            }
        }
      else
        {
          unsigned char        info;
          int        len;
          SCM        cpid;
          SCM        rport;
          SCM        wport;

          (void)close(p[1]);
          /*
           *        Synchronize with child process - it should send us a byte
           *        when everything is set up - immediately before the exec.
           */
          len = read(p[0], &info, 1);
          if (len != 1)
            {
              (void)close(p[0]);
              (void)close(master);
#if        HAVE_WAITPID
              {
                int        status;
                int        opts = 0;
                (void)waitpid(pid, &status, opts);
              }
#endif
              scm_misc_error("pty-child", "failed to sync with child",
                        SCM_EOL);
            }
          if (info == '\0')
            {
              (void)close(p[0]);
            }
          else
            {
              char        buf[1024];
              int        base;

              sprintf(buf, "child failed to open %s: ", slave_name);
              base = strlen(buf);
        
              len = read(p[0], &buf[base], (int)info);
              if (len > 0)
                {
                  buf[base+len] = '\0';
                }
              else
                {
                  strcat(buf, "unknown error");
                }
              (void)close(p[0]);
              scm_misc_error("pty-child", buf, SCM_EOL);
            }
          cpid = scm_from_int(pid);
          rport = scm_fdopen(scm_from_int(master), scm_from_locale_string("r"));
          wport = scm_fdopen(scm_from_int(master), scm_from_locale_string("w"));
          ans = scm_cons(rport, scm_cons(wport, scm_cons(cpid, SCM_EOL)));
        }
    }
  else
    {
      scm_misc_error("pty-child", "failed to get master pty", SCM_EOL);
    }
  return ans;
}
#undef FUNC_NAME

void
scm_init_greg()
{
#ifndef SCM_MAGIC_SNARFER
#include "libgreg.x"
#endif
}
